### DISCUSSION TITLE
GHO Subgraph Development

### DISCUSSION THREAD

**Discussion by defispartan** [aave/protocol-subgraphs#91] [General]
🧵  Open thread for GHO subgraph development

[GHO Subgraph](https://thegraph.com/hosted-service/subgraph/aave/protocol-v3-goerli-gho)

[Development Branch](https://github.com/aave/protocol-subgraphs/pull/92)

[Deployed GHO Contracts Goerli](https://docs.gho.xyz/developer-docs/contracts-overview)

### Current Schema

```
type Facilitator @entity {
  id: ID!
  bucketLevel: BigInt!
  bucketCapacity: BigInt!
  label: String!
  facilitatorLevelHistory: [FacilitatorLevelUpdated!]! @derivedFrom(field: "facilitator")
  facilitatorCapacityHistory: [FacilitatorCapacityUpdated!]! @derivedFrom(field: "facilitator")
  lifetimeFeesDistributedToTreasury: BigInt!
  treasuryDistributionHistory: [FacilitatorTreasuryDistribution!]!
    @derivedFrom(field: "facilitator")
  treasuryUpdateHistory: [FacilitatorTreasuryUpdated!]! @derivedFrom(field: "facilitator")
}

type FacilitatorLevelUpdated @entity {
  id: ID!
  txHash: Bytes!
  facilitator: Facilitator!
  oldBucketLevel: BigInt!
  newBucketLevel: BigInt!
}

type FacilitatorCapacityUpdated @entity {
  id: ID!
  txHash: Bytes!
  facilitator: Facilitator!
  oldBucketCapacity: BigInt!
  newBucketCapacity: BigInt!
}

type FacilitatorTreasuryDistribution @entity {
  id: ID!
  facilitator: Facilitator!
  amount: BigInt!
  treasury: Bytes!
  newLifetimeFeesDistributedToTreasury: BigInt!
  txHash: Bytes!
}

type FacilitatorTreasuryUpdated @entity {
  id: ID!
  facilitator: Facilitator!
  previousTreasury: Bytes!
  newTreasury: Bytes!
  txHash: Bytes!
}

type GhoFlashMinter @entity {
  id: ID!
  facilitator: Facilitator!
  fee: BigInt!
  maxFee: BigInt!
  ghoFlashMintHistory: [GhoFlashMint!]! @derivedFrom(field: "ghoFlashMinter")
  feeUpdates: [GhoFlashMinterFeeUpdate!]! @derivedFrom(field: "ghoFlashMinter")
}

type GhoFlashMint @entity {
  id: ID!
  ghoFlashMinter: GhoFlashMinter!
  receiver: Bytes!
  initiator: Bytes!
  amount: BigInt!
  fee: BigInt!
  timestamp: Int!
}

type GhoFlashMinterFeeUpdate @entity {
  id: ID!
  ghoFlashMinter: GhoFlashMinter!
  oldFee: BigInt!
  newFee: BigInt!
  txHash: Bytes!
}

type GhoDiscount @entity {
  id: ID!
  discountToken: Bytes!
  discountRate: BigInt!
  ghoDiscountedPerDiscountToken: BigInt!
  minDebtTokenBalance: BigInt!
  minDiscountTokenBalance: BigInt!
  ghoDiscountHistoryItems: [GhoDiscountHistoryItem!]! @derivedFrom(field: "ghoDiscount")
  discountTokenUpdates: [GhoDiscountTokenUpdate!]! @derivedFrom(field: "ghoDiscount")
}

type GhoDiscountHistoryItem @entity {
  id: ID!
  txHash: Bytes!
  ghoDiscount: GhoDiscount!
  oldGhoDiscountStrategy: Bytes!
  newGhoDiscountStrategy: Bytes!
  newDiscountRate: BigInt!
  newGhoDiscountedPerDiscountToken: BigInt!
  newMinDebtTokenBalance: BigInt!
  newMinDiscountTokenBalance: BigInt!
}

type GhoDiscountTokenUpdate @entity {
  id: ID!
  txHash: Bytes!
  ghoDiscount: GhoDiscount!
  oldDiscountToken: Bytes!
  newDiscountToken: Bytes!
}

```

### Events to index

IGHOToken.sol:
- [x] `event FacilitatorAdded(address indexed facilitatorAddress, bytes32 indexed label, uint256 bucketCapacity);`
- [x] `event FacilitatorRemoved(address indexed facilitatorAddress);`
- [x] `event FacilitatorBucketCapacityUpdated(address indexed facilitatorAddress, uint256 oldCapacity, uint256 newCapacity);`
- [x] `event FacilitatorBucketLevelUpdated(address indexed facilitatorAddress, uint256 oldLevel, uint256 newLevel);`

IGHOAToken.sol is IGhoFacilitator:
~~- [ ] `event VariableDebtTokenSet(address indexed variableDebtToken);`~~ handled in reserve initializer
- [x] `event FeesDistributedToTreasury(address indexed ghoTreasury, address indexed asset, uint256 amount);`
- [x] `event GhoTreasuryUpdated(address indexed oldGhoTreasury, address indexed newGhoTreasury);`

IGHOVariableDebtToken.sol:
~~- [ ] `event ATokenSet(address indexed aToken);`~~ handled in reserve initializer
- [x] `event DiscountRateStrategyUpdated(address indexed oldDiscountRateStrategy, address indexed newDiscountRateStrategy);`
- [x] `event DiscountTokenUpdated(address indexed oldDiscountToken, address indexed newDiscountToken);`
~~- [ ] `event DiscountPercentLocked(user, newDiscountPercent, newRebalanceTimestamp);`~~
~~- [ ] `event DiscountLockPeriodUpdated(uint256 indexed oldDiscountLockPeriod, uint256 indexed newDiscountLockPeriod);`~~

IGHOFlashMinter.sol:
- [x] `event FeeUpdated(uint256 oldFee, uint256 newFee);`
- [x] `event FlashMint(address indexed receiver, address indexed initiator, address asset, uint256 indexed amount, uint256 fee);`
- [x] `event FeesDistributedToTreasury(address indexed ghoTreasury, address indexed asset, uint256 amount);`
- [x] `event GhoTreasuryUpdated(address indexed oldGhoTreasury, address indexed newGhoTreasury);`

**Comment by defispartan** [comment]
Still dealing with one issue with the `FacilitatorAdded` event, the `label` is coming through as blank.

It's emitted in the FacilitatorAdded event: https://goerli.etherscan.io/tx/0x427be68e627e99e429d9465d2328ef8ce4fb9af215d4ecf56bb72e1f52af1543

And encoded here: https://github.com/aave/gho-core/blob/5ac3ed8d1fd5edea9ad3ee9b0942e3d211dec99a/src/contracts/gho/GhoToken.sol#L84

but I can't figure out how to decode it in the gho-token mapping

**Reply by ronlv10** [reply]
Maybe I'm missing something but I noticed that the code pointer you [shared](https://github.com/aave/gho-core/blob/5ac3ed8d1fd5edea9ad3ee9b0942e3d211dec99a/src/contracts/gho/GhoToken.sol#L84) uses `keccak256` (for the label) which is one-way hashing function.

Looking at the codebase I see that the code has changed 2 months ago ([here](https://github.com/aave/gho-core/commit/0bcf294b2758f50fb572e56efea4858224af22a6#diff-780f148250922baa88fdad3a7521e0a9128c2e27083f67bdb61218cf49ce1e0f)), so I think that the etherscan you shared is not updated with the new structure?  not sure if it's the reason you couldn't parse it, but IIUC it will not work for new facilitators or when moving to mainnet?

**Reply by defispartan** [reply]
Thanks for looking into this! Confirmed with contracts team that it's intended to be hashed, and because of this it's not possible to recover. They said that the previous version was being hashed as well, so this change just made it explicit.

I updated and re-deployed to just store the hashed hex string for label.

**Reply by defispartan** [reply]
Update on this, `label` is now stored in plaintext from calling `getFacilitor` in the `FacilitatorAdded` handler

**Comment by defispartan** [comment]
Created a separate [staging](https://thegraph.com/hosted-service/subgraph/aave/protocol-v3-goerli-gho-staging) deployment to keep the main endpoint stable while developing.

Staging deployment adds treasury entities, will deploy to main endpoint once it's synced and stable. Currently there is no treasury data coming through since Goerli treasury is [not configured](https://goerli.etherscan.io/address/0xbc249a1d91dbd54ceefbbdc66e01fc515fa3980c#readContract#F11)

**Comment by defispartan** [comment]
Re-deployed latest version of https://github.com/aave/protocol-subgraphs/pull/92 to https://thegraph.com/hosted-service/subgraph/aave/protocol-v3-goerli-gho. All events in initial schema are indexed ✔️
